server/src/gnome/prompter.rs

```diff
@@ -0,0 +1,450 @@
+            return Err(ServiceError::NoSuchObject(
+                "Prompt does not exist.".to_string(),
+            ));
+        };
+
+        let prompt_role = prompt.role();
+        match prompt_role {
+            PromptRole::Lock => {
+                self.prompt(
+                    prompt_role,
+                    reply,
+                    exchange,
+                    connection.clone(),
+                    prompt.path().clone(),
+                    prompt.objects().clone(),
+                )
+                .await?;
+            }
+            PromptRole::Unlock => {
+                self.prompt(
+                    prompt_role,
+                    reply,
+                    exchange,
+                    connection.clone(),
+                    prompt.path().clone(),
+                    prompt.objects().clone(),
+                )
+                .await?;
+            }
+            PromptRole::CreateCollection => todo!(),
+        };
+
+        Ok(())
+    }
+
+    pub async fn prompt_done(
+        &self,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
+    ) -> Result<(), ServiceError> {
+        if let Some(prompt) = self.service.prompt(&self.prompt_path).await {
+            let path = prompt.path();
+            object_server.remove::<Prompt, _>(path).await?;
+            self.service.remove_prompt(path).await;
+        }
+        object_server.remove::<Self, _>(&self.path).await?;
+
+        Ok(())
+    }
+}
+
+impl PrompterCallback {
+    pub async fn new(window_id: &str, service: Service, prompt_path: OwnedObjectPath) -> Self {
+        let index = service.prompt_index().await;
+        Self {
+            window_id: window_id.to_owned(),
+            secret_exchange_pub_key: Default::default(),
+            secret_exchange_aes_key: Default::default(),
+            path: OwnedObjectPath::try_from(format!("/org/gnome/keyring/Prompt/p{index}")).unwrap(),
+            service,
+            prompt_path,
+        }
+    }
+
+    pub fn path(&self) -> &OwnedObjectPath {
+        &self.path
+    }
+
+    pub async fn object_label(&self, objects: &Vec<OwnedObjectPath>) -> Option<String> {
+        for object in objects {
+            if let Some(collection) = self.service.collection_from_path(object).await {
+                return Some(collection.label().await);
+            }
+        }
+        None
+    }
+
+    pub async fn set_secret_exchange_pub_key(&self, pub_key: String) {
+        *self.secret_exchange_pub_key.write().await = pub_key;
+    }
+
+    pub async fn secret_exchange_pub_key(&self) -> String {
+        self.secret_exchange_pub_key.read().await.clone()
+    }
+
+    pub async fn set_secret_exchange_aes_key(&self, aes_key: String) {
+        *self.secret_exchange_aes_key.write().await = aes_key;
+    }
+
+    pub async fn secret_exchange_aes_key(&self) -> String {
+        self.secret_exchange_aes_key.read().await.clone()
+    }
+
+    pub async fn perform_prompt(
```

Comment:
All of these static methods have 0 usefulness at all. Drop all of them.

Reply:
Actually they do :) Consider `PrompterCallback::perform_prompt()` wrapper method. It's invoked in two places. So, this avoids code duplication in two places. And also with this approach I can avoid calling `.unwrap()` inside `tokio::spawn(async move` and avoid the following error, "^ cannot use the `?` operator in an async block that returns `()`"

Reply:
Well, you should make the spawned t ask return a Result instead of () https://docs.rs/tokio/latest/tokio/task/fn.spawn.html

Reply:
Thanks. I think we should keep these wrapper methods, they help a lot with code duplication.

Reply:
I haven't seen any code duplication so far.

Reply:
:) In `prompter.rs` file search for `stop_prompting`, `perform_prompt`, `prompt_completed` and `prompt`. Each of these static wrapper methods get executed twice.

Reply:
I think this is resolved too :)